Add @FieldOrder annotation so Rule classes are minimized

creativedrewy opened the pull request:
When digging through the Borsh operation, it became clear that for serializing/deserializing the fields need to be encountered in the same order so the bytes can be read/written properly. Your most recent Rule system really helped formalize making that happen, but I realized that for many field types if we could just specify the order, that would be enough.

Hence this PR - now you can just annotate the fields with ```@FieldOrder()``` and for many classes, no Rule class will be necessary. The rule system still works great for special cases, but hopefully this should reduce the boilerplate needed for most classes. Even the PublicKeyRule could likely be simplified/removed as a result of this annotation, though I didn't do that work for this PR.

creativedrewy commented on Solana/src/test/java/com/solana/models/MetaplexInfra.kt:
```diff
@@ -0,0 +1,18 @@
+import com.solana.vendor.borshj.FieldOrder
+
+data class MetaplexMeta(
+    @FieldOrder(0) val key: Byte,
```
The number in the order annotation is even sorta arbitrary, all that matters is that the numbers are from low to high alongside the relevant field.

creativedrewy commented on Solana/src/main/java/com/solana/vendor/borshj/BorshInput.kt:
```diff
@@ -46,17 +47,19 @@ interface BorshInput {
             val declaredFields = klass.declaredFields
-                .filter { kotlinParameters.map{ kf -> kf.name }.contains(it.name) }
+                .filter { kotlinParameters.map { kf -> kf.name }.contains(it.name) }
+                .sortedBy { field -> field.getAnnotation(FieldOrder::class.java).order }
```
I just realized that this doesn't allow for the annotation to be missing; I'm guessing if one of the fields doesn't have the annotation it will crash. The thing is, the field ordering is super important for borsh to work. So I'm not sure if this should be allowed to be optional.

ajamaica replied:
I totally got the point. We can ensure they  have e to be there.

creativedrewy replied:
Added a check in both in & out methods to check for any missing annotations. User-readable message is thrown if missing.

ajamaica reviewed (commented):
This is a welcome improvement. I  would consider putting the metaplex object in a different package name metaplex models inside the programs folder. This is not a native program. 

I think the tests are missing the field annotation that may be the reason of the failure tests.

creativedrewy commented:
> This is a welcome improvement. I  would consider putting the metaplex object in a different package name metaplex models inside the programs folder. This is not a native program. 
> 
> I think the tests are missing the field annotation that may be the reason of the failure tests.
> 
> 

The metaplex objects are in the test dir, do you want them in the actual library model objects? I was just using them to test various aspects of deserializing with the annotation.

I'll get the test check passing too!

creativedrewy commented:
You were right, I had failed to update the test classes with the annotation! I updated the tests but also updated the serialize/deserialize to throw an InvalidArgumentException if the annotation is missing on any classes.

ajamaica commented:
Leave the metaplex object on tests. That's my mistake

ajamaica commented:
We might add a sub lib that has support for all programs and objects for projects

creativedrewy commented:
> We might add a sub lib that has support for all programs and objects for projects

Yeah that sounds like a really good idea.
